bin/check-immutability.py

```diff
@@ -0,0 +1,54 @@
+#!/usr/bin/env python 
+
+import json
+import subprocess
+import sys
+
+oldf = sys.argv[1]
+newf = sys.argv[2]
+
+immutable_keys = ('property', 'input', 'expected')
+
+# Use jq to flatten the test data, and parse it
+old = json.loads(subprocess.run([f"jq -r '[.. | objects | select(.uuid != null)]' {oldf}"], stdout=subprocess.PIPE, shell=True).stdout.decode('utf-8'))
+new = json.loads(subprocess.run([f"jq -r '[.. | objects | select(.uuid != null)]' {newf}"], stdout=subprocess.PIPE, shell=True).stdout.decode('utf-8'))
```

Comment:
Requires `import shlex` (suggestion above).

If using stdout with `subprocess.run`, `capture_output=True` is recommended.

```suggestion
# Use jq to flatten the test data, and parse it
def get_test_data(filename):
	cmd = shlex.split(f"jq -r '[.. | objects | select(.uuid != null)]' {filename}")
    return json.loads(subprocess.run(cmd, capture_output=True, shell=True).stdout


old = get_test_data(oldf)
new = get_test_data(newf)
```

Reply:
Could you explain to me why `shlex.split` should be used here? The docs for it describe what it does but it's not clear to me why this should be used instead of calling the command directly.

Reply:
Really, there's no need to use `jq` to flatten the cases at all; it's prefectly possible (and even easier) in standard python:

```python3
def walk_cases(data):
    if "cases" in data:
        for case in data["cases"]:
            yield from walk_cases(case)
            if "uuid" in case and "description" in case:
               # if desired, yield case["uuid"]
                yield case
               
```

Reply:
@SaschaMann Depending on your input string, `subprocess.run` may interpret `jq -r ...` as one big command with whitespace in the filename, and then fail when it determines that no executable with that name exists in your path. Using `shlex.split` means that it gets parsed properly into an input list ahead of time.

Reply:
Thanks for the explanation. It didn't do that here, so I think shlex isn't needed in this particular case because the format of the inputs is well-known, while making it harder to follow/read for non-Python users. If it fails, that's a sign that the input is somehow wrong which means the commit also contains wrong data.

While I like that Python-only implementation, I think I'd rather go with the jq approach since we already use jq elsewhere in the repo and the Python-only solution introduces more complexity without, imo, much to gain. I'd rather keep a script like this as "dumb" as possible and stick to basic constructs whenever possible. The `subprocess.run` call is a mess but it's enough to understand that it runs the command and saves the output without having to understand the details.

I appreciate the suggestions though